src/classes/vehicle.py

```diff
@@ -87,73 +90,121 @@ def update_cmd(old_cmd: Command, t: np.array, a: np.array, elapsed_time: float=0
+        vehicle.velocity = 0
+        print("Invalid command: Distance to leading vehicle is too short")
+        return
+    
+    # required deceleration to stop car at fake distance away
+    required_deceleration = calculate_deceleration(final_velocity, initial_velocity, distance, safety_distance)
+
+    new_t = np.array([cur_time, cur_time + 0.01])
+    new_a = np.array([required_deceleration, leading_vehicle.acceleration])
+    vehicle.command = update_cmd(vehicle.command, new_t, new_a, cur_time)
+
+
+def handle_no_leading_vehicle(vehicle, cur_time: float) -> None:
+    """Handle the case where there is no leading vehicle."""
+    acceleration_distance = 10
+    initial_velocity = vehicle.velocity
+
+    if abs(vehicle.velocity - vehicle.default_velocity) > 0.01:
+        required_acceleration = (vehicle.default_velocity**2 - initial_velocity**2) / (2 * acceleration_distance)
+
+        new_t = np.array([cur_time, cur_time + 0.01])
+        new_a = np.array([required_acceleration, vehicle.acceleration])
+        vehicle.command = update_cmd(vehicle.command, new_t, new_a, cur_time)
+
+
+def check_traffic_lights(vehicle, cur_time: float) -> None:
+    """Check traffic lights and update commands based on their states."""
+    for r in vehicle.route.pos_to_edge_map:
+        if vehicle.route_position < r[1]:
+            edge = vehicle.route.pos_to_edge_map[r]
+
+            if edge.traffic_light and vehicle.leading_vehicle:
+                if should_ignore_light_due_to_leading_vehicle(vehicle, edge):
+                    continue
+
+            if edge.traffic_light:
+                handle_traffic_light(vehicle, edge, cur_time)
+
+
+def should_ignore_light_due_to_leading_vehicle(vehicle, edge) -> bool:
+    """Determine if the traffic light can be ignored due to the leading vehicle."""
+    distance_to_leading_vehicle = abs(vehicle.route_position - vehicle.leading_vehicle.route_position)
+    traffic_light_route_position = world_position_to_route_position(vehicle.route, edge, edge.traffic_light.node.position)
+    distance_to_traffic_light = abs(vehicle.route_position - traffic_light_route_position)
+
+    return distance_to_traffic_light > distance_to_leading_vehicle
+
+
+def handle_traffic_light(vehicle, edge, cur_time: float) -> None:
+    """Handle the traffic light's effect on the vehicle's command."""
+    initial_velocity = vehicle.velocity
+    traffic_light_state = get_light_state(edge.traffic_light)
+
+    if traffic_light_state != TrafficState.GREEN:
+        traffic_light_route_position = world_position_to_route_position(vehicle.route, edge, edge.traffic_light.node.position)
+        distance_to_traffic_light = abs(vehicle.route_position - traffic_light_route_position)
+        safety_distance = 9
+
+        if traffic_light_route_position < vehicle.route_position:
+            return
+        
+        if traffic_light_state == TrafficState.RED or (traffic_light_state == TrafficState.YELLOW and distance_to_traffic_light > 0):
```

Comment:
is there leeway distance happening here for yellow light? 0 is indicating to me that there isn't. Maybe I'm wrong...

Reply:
"distance_to_traffic_light = traffic_light_route_position - vehicle.route_position"

if distance_to_traffic_light < 0, this means the vehicle has already passed the traffic light

Reply:
If a car is 2 meters away from the traffic light and going fast, will there be leeway in that they will ignore the yellow light and just drive through?